File: go/types/type_test.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestTypes(t *testing.T) {
	assert := assert.New(t)
	vs := newTestValueStore()

	mapType := MakeMapType(StringType, FloaTType)
	setType := MakeSetType(StringType)
	mahType := MakeStructType("MahStruct",
		StructField{"Field1", StringType, false},
		StructField{"Field2", BoolType, false},
	)
	recType := MakeStructType("RecursiveStruct", StructField{"self", MakeCycleType("RecursiveStruct"), false})

	mRef := vs.WriteValue(mapType).TargetHash()
	setRef := vs.WriteValue(setType).TargetHash()
	mahRef := vs.WriteValue(mahType).TargetHash()
	recRef := vs.WriteValue(recType).TargetHash()

	assert.True(mapType.Equals(vs.ReadValue(mRef)))
	assert.True(setType.Equals(vs.ReadValue(setRef)))
	assert.True(mahType.Equals(vs.ReadValue(mahRef)))
	assert.True(recType.Equals(vs.ReadValue(recRef)))
}

func TestTypeType(t *testing.T) {
	assert.True(t, TypeOf(BoolType).Equals(TypeType))
}

func TestTypeRefDescribe(t *testing.T) {
	assert := assert.New(t)
	mapType := MakeMapType(StringType, FloaTType)
	setType := MakeSetType(StringType)

	assert.Equal("Bool", BoolType.Describe())
	assert.Equal("Float", FloaTType.Describe())
	assert.Equal("String", StringType.Describe())
	assert.Equal("UUID", UUIDType.Describe())
	assert.Equal("Int", IntType.Describe())
	assert.Equal("Uint", UintType.Describe())
	assert.Equal("Map<String, Float>", mapType.Describe())
	assert.Equal("Set<String>", setType.Describe())

	mahType := MakeStructType("MahStruct",
		StructField{"Field1", StringType, false},
		StructField{"Field2", BoolType, false},
	)
	assert.Equal("Struct MahStruct {\n  Field1: String,\n  Field2: Bool,\n}", mahType.Describe())
}

func TestTypeOrdered(t *testing.T) {
	assert := assert.New(t)
	assert.True(isKindOrderedByValue(BoolType.TargetKind()))
	assert.True(isKindOrderedByValue(FloaTType.TargetKind()))
	assert.True(isKindOrderedByValue(UUIDType.TargetKind()))
	assert.True(isKindOrderedByValue(StringType.TargetKind()))
	assert.True(isKindOrderedByValue(IntType.TargetKind()))
	assert.True(isKindOrderedByValue(UintType.TargetKind()))
	assert.True(isKindOrderedByValue(TupleKind))

	assert.False(isKindOrderedByValue(BlobType.TargetKind()))
	assert.False(isKindOrderedByValue(ValueType.TargetKind()))
	assert.False(isKindOrderedByValue(MakeListType(StringType).TargetKind()))
	assert.False(isKindOrderedByValue(MakeSetType(StringType).TargetKind()))
	assert.False(isKindOrderedByValue(MakeMapType(StringType, ValueType).TargetKind()))
	assert.False(isKindOrderedByValue(MakeRefType(StringType).TargetKind()))
}

func TestFlattenUnionTypes(t *testing.T) {
	assert := assert.New(t)
	assert.Equal(BoolType, MakeUnionType(BoolType))
	assert.Equal(MakeUnionType(), MakeUnionType())
	assert.Equal(MakeUnionType(BoolType, StringType), MakeUnionType(BoolType, MakeUnionType(StringType)))
	assert.Equal(MakeUnionType(BoolType, StringType, FloaTType), MakeUnionType(BoolType, MakeUnionType(StringType, FloaTType)))
	assert.Equal(BoolType, MakeUnionType(BoolType, BoolType))
	assert.Equal(BoolType, MakeUnionType(BoolType, MakeUnionType()))
	assert.Equal(BoolType, MakeUnionType(MakeUnionType(), BoolType))
	assert.True(MakeUnionType(MakeUnionType(), MakeUnionType()).Equals(MakeUnionType()))
	assert.Equal(MakeUnionType(BoolType, FloaTType), MakeUnionType(BoolType, FloaTType))
	assert.Equal(MakeUnionType(BoolType, FloaTType), MakeUnionType(FloaTType, BoolType))
	assert.Equal(MakeUnionType(BoolType, FloaTType), MakeUnionType(BoolType, FloaTType, BoolType))
	assert.Equal(MakeUnionType(BoolType, FloaTType), MakeUnionType(MakeUnionType(BoolType, FloaTType), FloaTType, BoolType))
}

func TestVerifyStructFieldName(t *testing.T) {
	assert := assert.New(t)

	assertInvalid := func(n string) {
		assert.Panics(func() {
			MakeStructType("S", StructField{n, StringType, false})
		})
	}
	assertInvalid("")
	assertInvalid(" ")
	assertInvalid(" a")
	assertInvalid("a ")
	assertInvalid("0")
	assertInvalid("_")
	assertInvalid("0a")
	assertInvalid("_a")
	assertInvalid("💩")

	assertValid := func(n string) {
		MakeStructType("S", StructField{n, StringType, false})
	}
	assertValid("a")
	assertValid("A")
	assertValid("a0")
	assertValid("a_")
	assertValid("a0_")
}

func TestVerifyStructName(t *testing.T) {
	assert := assert.New(t)

	assertInvalid := func(n string) {
		assert.Panics(func() {
			MakeStructType(n)
		})
	}

	assertInvalid(" ")
	assertInvalid(" a")
	assertInvalid("a ")
	assertInvalid("0")
	assertInvalid("_")
	assertInvalid("0a")
	assertInvalid("_a")
	assertInvalid("💩")

	assertValid := func(n string) {
		MakeStructType(n)
	}
	assertValid("")
	assertValid("a")
	assertValid("A")
	assertValid("a0")
	assertValid("a_")
	assertValid("a0_")
}

func TestStructUnionWithCycles(tt *testing.T) {
	inodeType := MakeStructTypeFromFields("Inode", FieldMap{
		"attr": MakeStructTypeFromFields("Attr", FieldMap{
			"ctime": FloaTType,
			"mode":  FloaTType,
			"mtime": FloaTType,
		}),
		"contents": MakeUnionType(
			MakeStructTypeFromFields("Directory", FieldMap{
				"entries": MakeMapType(StringType, MakeCycleType("Inode")),
			}),
			MakeStructTypeFromFields("File", FieldMap{
				"data": BlobType,
			}),
			MakeStructTypeFromFields("Symlink", FieldMap{
				"targetPath": StringType,
			}),
		),
	})

	t1, _ := inodeType.Desc.(StructDesc).Field("contents")
	t2 := DecodeValue(EncodeValue(t1), nil)

	assert.True(tt, t1.Equals(t2))
	// Note that we cannot ensure pointer equality between t1 and t2 because the
	// types used to the construct the Unions, while eventually equivalent, are
	// not identical due to the potentially differing placement of the Cycle type.
	// We do not remake Union types after putting their component types into
	// their canonical ordering.
}

func TestHasStructCycles(tt *testing.T) {
	assert := assert.New(tt)

	assert.False(HasStructCycles(BoolType))
	assert.False(HasStructCycles(BlobType))
	assert.False(HasStructCycles(FloaTType))
	assert.False(HasStructCycles(StringType))
	assert.False(HasStructCycles(TypeType))
	assert.False(HasStructCycles(ValueType))
	assert.Panics(func() {
		HasStructCycles(MakeCycleType("Abc"))
	})

	assert.False(HasStructCycles(MakeStructType("")))
	assert.False(HasStructCycles(MakeStructType("A")))

	assert.True(HasStructCycles(
		MakeStructType("A", StructField{"a", MakeStructType("A"), false})))
	assert.True(HasStructCycles(
		MakeStructType("A", StructField{"a", MakeCycleType("A"), false})))
	assert.True(HasStructCycles(
		MakeSetType(MakeStructType("A", StructField{"a", MakeCycleType("A"), false}))))
	assert.True(HasStructCycles(
		MakeStructType("A", StructField{"a", MakeSetType(MakeCycleType("A")), false})))

	assert.False(HasStructCycles(
		MakeMapType(
			MakeStructType("A"),
			MakeStructType("A"),
		),
	))
	assert.False(HasStructCycles(
		MakeMapType(
			MakeStructType("A"),
			MakeCycleType("A"),
		),
	))

	assert.False(HasStructCycles(
		MakeStructType("",
			StructField{"a", MakeStructType("",
				StructField{"b", BoolType, false},
			), false},
			StructField{"b", MakeStructType("",
				StructField{"b", BoolType, false},
			), false},
		)),
	)
}

File: js/noms/src/type-test.js
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

// @flow

import {TestDatabase} from './test-util.js';
import {assert} from 'chai';
import {
  blobType,
  boolType,
  makeCycleType,
  makeMapType,
  makeSetType,
  makeStructType,
  makeUnionType,
  numberType,
  stringType,
  typeType,
  getTypeOfValue,
} from './type.js';
import type {Type} from './type.js';
import {suite, test} from 'mocha';
import {equals} from './compare.js';
import {encodeValue, decodeValue} from './codec.js';
import {notNull} from './assert.js';

suite('Type', () => {
  test('types', async () => {
    const db = new TestDatabase();

    const mapType = makeMapType(stringType, numberType);
    const setType = makeSetType(stringType);
    const mahType = makeStructType('MahStruct', {
      Field1: stringType,
      Field2: boolType,
    });
    const mapRef = db.writeValue(mapType).targetHash;
    const setRef = db.writeValue(setType).targetHash;
    const mahRef = db.writeValue(mahType).targetHash;

    assert.isTrue(equals(mapType, await db.readValue(mapRef)));
    assert.isTrue(equals(setType, await db.readValue(setRef)));
    assert.isTrue(equals(mahType, await db.readValue(mahRef)));
    await db.close();
  });

  test('type Type', () => {
    assert.isTrue(equals(boolType.type, typeType));
  });

  test('getTypeOfValue', () => {
    assert.equal(boolType, getTypeOfValue(true));
    assert.equal(boolType, getTypeOfValue(false));
    assert.equal(numberType, getTypeOfValue(42));
    assert.equal(numberType, getTypeOfValue(0));
    assert.equal(stringType, getTypeOfValue('abc'));
    assert.equal(typeType, getTypeOfValue(stringType));
  });

  test('flatten union types', () => {
    assert.equal(makeUnionType([boolType]), boolType);
    assert.deepEqual(makeUnionType([]), makeUnionType([]));
    assert.deepEqual(makeUnionType([boolType, makeUnionType([stringType])]),
                     makeUnionType([boolType, stringType]));
    assert.deepEqual(makeUnionType([boolType, makeUnionType([stringType, numberType])]),
                     makeUnionType([boolType, stringType, numberType]));
    assert.equal(makeUnionType([boolType, boolType]), boolType);
    assert.equal(makeUnionType([boolType, makeUnionType([])]), boolType);
    assert.equal(makeUnionType([makeUnionType([]), boolType]), boolType);
    assert.isTrue(equals(makeUnionType([makeUnionType([]), makeUnionType([])]), makeUnionType([])));
    assert.deepEqual(makeUnionType([boolType, numberType]), makeUnionType([boolType, numberType]));
    assert.deepEqual(makeUnionType([numberType, boolType]), makeUnionType([boolType, numberType]));
    assert.deepEqual(makeUnionType([boolType, numberType, boolType]),
                     makeUnionType([boolType, numberType]));
    assert.deepEqual(makeUnionType([makeUnionType([boolType, numberType]), numberType, boolType]),
                     makeUnionType([boolType, numberType]));
  });

  test('verify struct field name', () => {
    function assertInvalid(n: string) {
      assert.throw(() => {
        makeStructType('S', {[n]: stringType});
      });
    }
    assertInvalid('');
    assertInvalid(' ');
    assertInvalid(' a');
    assertInvalid('a ');
    assertInvalid('0');
    assertInvalid('_');
    assertInvalid('0a');
    assertInvalid('_a');
    assertInvalid('💩');

    function assertValid(n: string) {
      makeStructType('S', {[n]: stringType});
    }
    assertValid('a');
    assertValid('A');
    assertValid('a0');
    assertValid('a_');
    assertValid('a0_');
  });

  test('verify struct name', () => {
    function assertInvalid(n: string) {
      assert.throw(() => {
        makeStructType(n, {});
      });
    }
    assertInvalid(' ');
    assertInvalid(' a');
    assertInvalid('a ');
    assertInvalid('0');
    assertInvalid('_');
    assertInvalid('0a');
    assertInvalid('_a');
    assertInvalid('💩');

    function assertValid(n: string) {
      makeStructType(n, {});
    }
    assertValid('');
    assertValid('a');
    assertValid('A');
    assertValid('a0');
    assertValid('a_');
    assertValid('a0_');
  });

  test('type.describe', () => {
    // This is tested exaustively as part of HRS, just testing here that
    // type.describe() is present and works.
    [
      [numberType, 'Number'],
      [stringType, 'String'],
      [makeSetType(numberType), 'Set<Number>'],
    ].forEach(([t, desc]: [Type<any>, string]) => {
      assert.equal(t.describe(), desc);
    });
  });

  test('union with cycle', () => {
    const inodeType = makeStructType('Inode', {
      attr: makeStructType('Attr', {
        ctime: numberType,
        mode: numberType,
        mtime: numberType,
      }),
      contents: makeUnionType([
        makeStructType('Directory', {
          entries: makeMapType(stringType, makeCycleType(1)),
        }),
        makeStructType('File', {
          data: blobType,
        }),
        makeStructType('Symlink', {
          targetPath: stringType,
        }),
      ]),
    });

    const vr: any = null;
    const t1 = notNull(inodeType.desc.getField('contents'));
    const t2 = decodeValue(encodeValue(t1, null), vr);
    assert.isTrue(equals(t1, t2));
    /*
     * Note that we cannot ensure pointer equality between t1 and t2 because the types used to the
     * construct the Unions, while eventually equivalent, are not identical due to the potentially
     * differing placement of the Cycle type. We do not remake Union types after putting their
     * component types into their canonical ordering.
     */
  });
});
